### DISCUSSION TITLE
How to save the json config trigger value in a python variable ?

### DISCUSSION THREAD

**Discussion by eikesf** [apache/airflow#24714] [Q&A]
I'm trying to trigger my dag with config and in there i'm passing `{"estado":"ES"}` what a need to do is to take this value and save it in a python variable to use inside a python function, how i do this ?. I've seen some tutorials saying that the command is `{{ dag_run.conf['estado'] }}` but it only works when executing with a bash operator, also i tried to execute python subprocess but it didn't work. 
Below is the image of my trigger dag w/ config in the airflow UI
![image](https://user-images.githubusercontent.com/88998110/176267649-96ce5a4f-0bdb-4b3b-999c-839106c2ff8f.png)

Below is some of the ways that i tried to grab it

```
def get_estado(**kwargs):
        #estado = kwargs["estado"]
        #estado = dag_run.conf['estado']
        #estado={{dag_run.conf.get('estado')}}
        #estado = os.system("{{ dag_run.conf['estado'] }}")
        #bashcmd = "{{ dag_run.conf['estado'] }}"
        #process = subprocess.Popen(bashcmd, stdout=subprocess.PIPE)
        #estado, error = process.communicate()
        #estado = "PR"
        if(estado == 'ES'):
            nome_estado='ES'
            ip_estado='http://0.0.0.0'
        elif(estado == 'PR'):
            nome_estado='PR'
            ip_estado='http://1.1.1.1'
        elif(estado == 'RJ'):
            nome_estado='RJ'
            ip_estado=''
        elif(estado == 'SC'):
            nome_estado='SC'
            ip_estado=''
        elif(estado == 'SP'):
            nome_estado='SP'
            ip_estado=''
        else:
            nome_estado='Estado não cadastrado'
            ip_estado='O IP deste estado não foi cadastrado'
        return nome_estado, ip_estado
```

**Comment by potiuk** [comment]
You can only pass it via 

https://airflow.apache.org/docs/apache-airflow/stable/dag-run.html?highlight=dag_run

> Note: The parameters from dag_run.conf can only be used in a template field of an operator.

You have to add `op_args` or `op_kwargs` in your PythonOperator and pass it there.

**Reply by eikesf** [reply]
Potiuk i've tried to run my function as a example that i found in your profile. My function is like this know:

```
@task(task_id="run_this")
def get_estado(estado=None):   
    if(estado == "ES"):
        nome_estado="ES"
        ip_estado="http://1.1.1.1"
        print(f"Estado {nome_estado} com ip {ip_estado}")
    elif(estado == "PR"):
        nome_estado="PR"
        ip_estado="http://2.2.2.2"
        print(f"Estado {nome_estado} com ip {ip_estado}")
    elif(estado == "RJ"):
        nome_estado="RJ"
        ip_estado="http://3.3.3.3"
        print(f"Estado {nome_estado} com ip {ip_estado}")
    elif(estado == "SC"):
        nome_estado='SC'
        ip_estado="http://4.4.4.4"
        print(f"Estado {nome_estado} com ip {ip_estado}")
        ip_estado=""
    elif(estado == "SP"):
        nome_estado="SP"
        ip_estado="http://5.5.5.5"
        print(f"Estado {nome_estado} com ip {ip_estado}")
    else:
        nome_estado="Estado {} não cadastrado".format(estado)
        ip_estado="O IP do estado {} não foi cadastrado. tipo de dado {}, ".format(estado, type(estado))
    return nome_estado,ip_estado

with DAG(dag_id="PROPESQ_Anon", default_args=default_args, schedule_interval="@Weekly", start_date=datetime(2022,6,1), catchup=False, tags=["Propesq","anonimizar"]) as dag:
    hoje  = date.today()
    ontem = hoje - timedelta(days=1)
    ano = ontem.strftime("%Y")
    mes = ontem.strftime("%m")
    dia = ontem.strftime("%d")
    ontem = ano+mes+dia

    get_estado = get_estado(estado="{{dag_run.conf.get('estado')}}")

    baixa_backup = BashOperator(
        task_id = "baixar_backup",
        bash_command = "cd /opt/airflow/backups && wget --user xxx--password '1234' {}:81/DATABASE/PGBACKUP-{}.tar.bz2".format(ip_estado, ontem),
        dag = dag
    ) 
```
With this i run my dag with config like the previous one and i got the answer that i was expecting.
`[2022-07-05, 11:13:12 -03] {logging_mixin.py:115} INFO - Estado ES com ip http://1.1.1.1/)`
`[2022-07-05, 11:13:12 -03] {python.py:173} INFO - Done. Returned value was: ('ES', 'http://1.1.1.1/')`

What i need is a way to grab this return statement and use in the other tasks from my dag such as `baixa_backup` to do this i need to create an Xcom ? What's the right way to grab this return and use in the other tasks?

Thanks for the reply

**Reply by potiuk** [reply]
Yes. XCom.

**Reply by eikesf** [reply]
Do you have an example on how can i grab multiple values from one Xcom just like the one in my last comment ('ES', 'http://1.1.1.1/') ?  i've tried to use by the index and it didn't work. What's the correct way to do it?

**Reply by potiuk** [reply]
Find it in the docs (it's easy).

**Reply by potiuk** [reply]
You can also find a lot of examples in the source code of airflow - just search for xcom in both cases. This way you will learn much better than by getting ready answer.

**Reply by eikesf** [reply]
I'll try to find and understant it. Thanks potiuk
